project/apps/web/components/history/history-table/HistoryTable.tsx

```diff
@@ -0,0 +1,180 @@
+import { HistoryTableToolbar } from './HistoryTableToolbar';
+
+export type CollabInfoiWithPartner = CollabInfoDto & {
+  partner: {
+    id: string;
+    username: string;
+  };
+};
+
+export function HistoryTable() {
+  const { userData } = useMe();
+  const { confirmLoading, setConfirmLoading } = useQuestionsStore();
+
+  const [pagination, setPagination] = useState<PaginationState>({
+    pageIndex: 0,
+    pageSize: 10,
+  });
+
+  const resetPagination = () => {
+    setPagination((prev) => ({
+      ...prev,
+      pageIndex: 0,
+    }));
+  };
+
+  const [sorting, setSorting] = useState<SortingState>([
+    {
+      id: 'started_at',
+      desc: true,
+    },
+  ]);
+
+  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>([]);
+  const debouncedColumnFilters = useDebounce(
+    columnFilters,
+    100,
+    () => setConfirmLoading(true),
+    () => setConfirmLoading(false),
+  );
+
+  useEffect(() => {
+    if (confirmLoading) {
+      setConfirmLoading(true);
+    } else {
+      setConfirmLoading(false);
+    }
+  }, [pagination, sorting, debouncedColumnFilters, setConfirmLoading]);
+
+  const { data } = useSuspenseQuery<CollabCollectionDto>({
+    queryKey: [QUERY_KEYS.Collab, pagination, sorting, debouncedColumnFilters],
+    queryFn: async () => {
+      const user_id = userData!.id;
+
+      const q_title = debouncedColumnFilters.find(
+        (f) => f.id === 'question.q_title',
+      )?.value as string;
+
+      const q_category = debouncedColumnFilters.find(
+        (f) => f.id === 'question.q_category',
+      )?.value as CATEGORY[];
+
+      const q_complexity = debouncedColumnFilters.find(
+        (f) => f.id === 'question.q_complexity',
+      )?.value as COMPLEXITY[];
+
+      const offset = pagination.pageIndex * pagination.pageSize;
+      const limit = pagination.pageSize;
+
+      const sort = sorting.map(
+        (s) =>
+          ({
+            field: s.id,
+            order: s.desc ? 'desc' : 'asc',
+          }) as SortQuestionsQueryDto,
+      );
+
+      const queryParams: CollabFiltersDto = {
+        user_id,
+        has_ended: true,
+        q_title,
+        q_category,
+        q_complexity,
+        offset,
+        limit,
+        sort,
+      };
+
+      return await fetchCollabs(queryParams);
+    },
+  });
+
+  const onPaginationChange = (updater: Updater<PaginationState>) => {
+    startTransition(() => {
+      setPagination(updater);
+    });
+  };
+
+  const onSortingChange = (updater: Updater<SortingState>) => {
+    startTransition(() => {
+      setSorting(updater);
+      resetPagination();
+    });
+  };
+
+  const onColumnFiltersChange = (updater: Updater<ColumnFiltersState>) => {
+    startTransition(() => {
+      setColumnFilters(updater);
+      resetPagination();
+    });
+  };
+
+  //  dataWithPartner includes a new 'partner' field, which is the other user in the collaboration
+  const dataWithPartner = data?.collaborations.map((collab) => {
```

Comment:
is it possible/suitable to memoize this perhaps? 
```
const dataWithPartner = useMemo(() => {
  if (!data || !userData) return [];
  return data.collaborations.map((collab) => {
    const partner = collab.collab_user1.id === userData.id
      ? collab.collab_user2
      : collab.collab_user1;
    return { ...collab, partner };
  }) as CollabInfoWithPartner[];
}, [data, userData]);

```

Reply:
thx for the suggestion 👍